---
title: "Lab 10: Population genomics based on high throughput sequencing (HTS)"
author: "Melise Edwards"
date: "11/28/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Lab 10 

https://jeffreyblanchard.github.io/EvoGeno/EvoGeno_Lab10_SNP.html


### Part I: Reading VCF Data 
(VCF: variant call format, preferred file format obtained from genome sequencing)


```{r}
library(vcfR)

```

```{r}
data(vcfR_example) # loads the objects 'gff', 'dna' and 'vcf' 
vcf
```
```{r}
strwrap(vcf@meta[1:7])
```

Note: The first line contains the version of the VCF format used in the file. This line is required. The second line specifies the software which created the VCF file. This is not required, so not all VCF files include it. When they do, the file becomes self documenting.  Some VCF files may contain a line for every chromosome (or supercontig or contig depending on your genome), so they may become rather long. Here, the remaining lines contain INFO and FORMAT specifications which define abbreviations used in the fix and gt portions of the file.


```{r}
queryMETA(vcf)
```
When the function queryMETA() is called with only a vcfR object as a parameter, it attempts to summarize the meta information. Not all of the information is returned. For example, ‘contig’ elements are not returned. This is an attempt to summarize information that may be most useful for comprehension of the file’s contents.

```{r}
queryMETA(vcf, element = 'DP')
```

```{r}
queryMETA(vcf, element = 'FORMAT=<ID=DP')
```
Here we’ve isolated the definition of ‘DP’ as a ‘FORMAT’ element. Note that the function queryMETA() includes the parameter nice which by default is TRUE and attempts to present the data in a nicely formatted manner. However, our query is performed on the actual information in the ‘meta’ region. It is therefore sometimes appropriate to set nice = FALSE so that we can see the raw data. In the above example the angled bracket (‘<’) is omitted from the nice = TRUE representation but is essential to distinguishing the ‘FORMAT’ element from the ‘INFO’ element.

*****

### Fix Region 

The fix region contains information for each variant which is sometimes summarized over all samples. The first eight columns of the fixed region are titled CHROM, POS, ID, REF, ALT, QUAL, FILTER and INFO. This is per variant information which is ‘fixed’, or the same, over all samples. The first two columns indicate the location of the variant by chromosome and position within that chromosome. Here, the ID field has not been used, so it consists of missing data (NA). The REF and ALT columns indicate the reference and alternate allelic states for a diploid sample. When multiple alternate allelic states are present they are delimited with commas. The QUAL column attempts to summarize the quality of each variant over all samples. The FILTER field is not used here but could contain information on whether a variant has passed some form of quality assessment. The eigth column, titled INFO, is a semicolon delimited list of information. It can be rather long and cumbersome. The function getFIX() will suppress this column by default.

```{r}
head(getFIX(vcf))
```
******

### Genotype region (gt region)

The gt (genotype) region contains information about each variant for each sample. The definition of these acronyms can be referenced by querying the the meta region.

```{r}
vcf@gt[1:6, 1:4]
```

```{r}
#lets query the meta data to see a definition for another acronym, for practice 
queryMETA(vcf, element = 'FORMAT=<ID=DP') # read depth
```
```{r}
queryMETA(vcf, element = 'FORMAT=<ID=AD') # allelic depths
```
```{r}
queryMETA(vcf, element = 'FORMAT=<ID=GT') # genotype

queryMETA(vcf, element = 'FORMAT=<ID=PL') # phred scaled likelihoods for genotypes 

queryMETA(vcf, element = 'FORMAT=<ID=GQ') # genotype quality 
```

*****

### vcfR

Using the R package vcfR, we can read VCF format files into memory using the function read.vcfR(). Once in memory we can use the head() method to summarize the information in the three VCF regions.

```{r}
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")

```

```{r}
head(vcf)
```

After we have made any manipulations of the file we can save it as a VCF file with the function write.vcf().

```{r}
write.vcf(vcf, "myVCFdata_filtered.vcf.gz")

```

*********


# Exercises Part I 

1. To find more information about 'read.vcfR', we would use the ? before the function. 

2. To learn what the acronym 'AD' stands for, we can use the code below:

```{r}
queryMETA(vcf, element = 'FORMAT=<ID=AD') # allelic depths 
```
3. To view the last lines of our 'fix' data, we would use the 'tail' function below:

```{r}
tail(vcf@fix)
```

 4. Qual is useful because it shows us the quality, see plot below:
 
```{r}
plot(vcf)
```
 
 5. To query sample names, use:
 
```{r}
colnames(vcf@gt)
```
 
 ******
 
# Part II - Analysis of Genome Data 
Examples of how genomic data can be analyzed.

### Opening data and viewing it 
 
```{r}
library('vcfR')
```

```{r}
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz")
```

```{r}
vcf
```
 *****
 
#### Converting VCF data to genlight object
The R package adegenet is a popular R package used for population genetic analysis and it works on data structures called ‘genlight’ objects.

```{r}
library("adegenet")
```


```{r}
x <- vcfR2genlight(vcf) #converting into a 'genlight' object
```
```{r}
x
```

A genlight object only supports biallelic, or binary, variants. That is, variants with no more than two alleles. However, variant call format data can include multiple alleles.

 In VCF data the alleles are delimited by either a pipe or a forward slash (‘|’, ‘/’ respectively). Because genlight objects only use biallelic loci the genotypes can be recoded as 0, 1 and 2. These correspond to homozygous for the reference or zero allele, heterozygote or homozygous for the first alternate allele.
 
We can validate this by checking a few select genotypes from both the vcfR object and the genlight object.

```{r}
# vcfR
gt <- extract.gt(vcf, element = "GT")
gt[c(2,6,18), 1:10]
```

```{r}
# genlight
t(as.matrix(x))[c(1,5,17), 1:3]
```

Samples are in rows while the variants are in columns. 


Yet another difference among VCF data and genlight objects is that in VCF data there is no concept of ‘population.’ The package adegenet was designed specifically for the analysis of population data, so its genlight object has a place (a ‘slot’) to hold this information. Because there is no population data in VCF data, if we want population data we’ll have to set it ourselves.


```{r}
pop(x) <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
popNames(x) # By using the as.factor() function we transform the “vector” into a “factor”
```
By using the as.factor() function we transform the “vector” into a “factor”. A factor understands that all of the elements that are named “us” or “eu” are all part of the same group.

```{r}
ploidy(x) <- 2 # ploidy = the number of sets of chromosomes in a cell, or in the cells of an organism.

```

*****

### Distance matrices 

```{r}
x.dist <- dist(x) # create a distance matrix from a genlight object
```


```{r}
x.dist <- poppr::bitwise.dist(x) # use package 'poppr'
```


**Terminology**
* SNP: single nucleotide polymorphism, substitution of a single nucleotide at a specific position in the genome. 
* indel: insertion or deletion of bases in the genome of an organism 

 The R package vcfR provides a link between VCF data and the R environment and it includes a simple genome browser to help visualize the effect of manipulations. Here we explore how we can use vcfR to survey genomic data for interesting features.


### Creating chromR objects


```{r}
# if not already in use, make sure to start this workflow with: >library(vcfR)

vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = "pinfsc50")
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = "pinfsc50")
gff_file <- system.file("extdata", "pinf_sc50.gff", package = "pinfsc50")

```

```{r}
# Input the files.
vcf <- read.vcfR(vcf_file, verbose = FALSE)
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")
```

```{r}
# Create a chromR object.
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=TRUE)
```

```{r}
chrom
```

There at least two ways to graphically view the chromR object. The first is plot() which plots histograms of some of data summaries.


```{r}
plot(chrom)
```
**More Terminology from Jaffe et al 2003**
* This assembly process produces sequence units (called **supercontigs** or scaffolds), which under the best of circumstances can approximate chromosomes, but which in general are smaller, less contiguous, and have errors. Within a supercontig, contiguous segments (or contigs) are punctuated by gaps (captured by paired reads) whose sizes are approximately known (as a consequence of knowing the insert lengths). The resulting draft sequence may be an end goal or may be the starting point for production of finished sequence, through additional clone-based sequencing.
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC430950/

```{r}
chromoqc(chrom, dp.alpha = 66) # called a chromo plot and distributes data along its chromosomal coordinates
```

### Processing chromR objects

```{r}
chrom <- proc.chromR(chrom, verbose = TRUE)
```

```{r}
plot(chrom) #identical to the one before, but now we have variant densities 
# When we observe the chromoqc plot we see that we now have variant densities, nucleotide content as well as a representation of where in our reference we have nucleotides (A, C, G or T) or where we have ambiguous nucleotides.
```

```{r}
chromoqc(chrom, dp.alpha = 66) #visualizing raw data that has come from a variant caller and other automated sources
```

In our section on quality control we presented methods on how to filter variants on various parameters as an attempt to omit low quality variants. We can use this data to create a chromR object and compare it to the above data.


```{r}
#vcf <- read.vcfR("pinfsc50_qc.vcf.gz", verbose = FALSE)
vcf <- read.vcfR("data/pinfsc50_filtered.vcf.gz", verbose = FALSE) # filtered data, steps described elsewhere for this 
chrom <- create.chromR(name="Supercontig", vcf=vcf, seq=dna, ann=gff, verbose=FALSE)
chrom <- proc.chromR(chrom, verbose = FALSE)
chromoqc(chrom, dp.alpha = 66) # these quality control steps filter out data and leave us with a smaller quantity of data 
```

*******

### Tabular summaries 

```{r}
head(chrom@var.info)
```


```{r}
head(chrom@win.info)
```

Note from Dr. Blanchard: While loading entire genomes into memory may not be practical due to resource limitations, it is frequently practical to break a genome up into fractions that can be processed given the resources available on any system. By processing a genome by chromosomes, or some other fraction, and saving this tabular data to file you can perform genome scans in an attempt to identify interesting features.

****

### Genetic Differentiation 

* "To address the question of within population diversity geneticists typically report heterozygosity. This is the probability that two alleles randomly chosen from a population will be different (Nei 1973)."

* In vcfR, the function genetic_diff() was implemented to measure population diversity and differentiation.

```{r}
data(vcfR_example)
pop <- as.factor(c("us", "eu", "us", "af", "eu", "us", "mx", "eu", "eu", "sa", "mx", "sa", "us", "sa", "Pmir", "us", "eu", "eu"))
myDiff <- genetic_diff(vcf, pops = pop, method = 'nei')
knitr::kable(head(myDiff[,1:15]))
```
The function returns the chromosome and position of each variant as provided in the VCF data. This should allow you to align its output with the VCF data. The heterozygosities for each population are reported as well as the total heterozygosity, followed by the number of alleles observed in each population. Note that in some populations zero alleles were observed. Populations with zero alleles reported heterozygosities of ‘NaN’ because of this absence of data.

```{r}
knitr::kable(head(myDiff[,16:19]))
```



```{r}
knitr::kable(round(colMeans(myDiff[,c(3:9,16,19)], na.rm = TRUE), digits = 3))
```

```{r}
#summarize using violin plots
library(ggplot2)
library(reshape2)
dpf <- melt(myDiff[,c(3:8,19)], varnames=c('Index', 'Sample'), value.name = 'Depth', na.rm=TRUE)

```

```{r}
p <- ggplot(dpf, aes(x=variable, y=Depth)) + geom_violin(fill="#2ca25f", adjust = 1.2)
p <- p + xlab("")
p <- p + ylab("")
p <- p + theme_bw()
p
```

### Exercises Part II 

```{r}
# 1. Gprimest by POS
plot(getPOS(vcf), myDiff$Gprimest,  pch = 20, col = "#1E90FF44", xlab = "", ylab = "", ylim = c(0, 1), xaxt = "n")
axis(side = 1, at = seq(0, 1e5, by = 1e4), labels = seq(0, 100, by = 10))
title(xlab='Genomic position (Kbp)')
title(ylab = expression(italic("G'"["ST"])))

```

```{r}
#2. small sample size?

table(pop)
```

```{r}
#3. zooming in on specific chromosome 

chromoqc(chrom, dp.alpha = 66, xlim = c(2e05, 4e05))
```

```{r}
#4. how to look for other data that might be of interest

queryMETA(vcf)
```

## EvoGeno References by Dr. Blanchard 

Danecek, Petr, Adam Auton, Goncalo Abecasis, Cornelis A Albers, Eric Banks, Mark A DePristo, Robert E Handsaker, et al. 2011. “The Variant Call Format and VCFtools.” Bioinformatics 27 (15): 2156–8. https://doi.org/10.1093/bioinformatics/btr330.

Grünwald, Niklaus J, Bruce A McDonald, and Michael G Milgroom. 2016. “Population Genomics of Fungal and Oomycete Pathogens.” Annual Review of Phytopathology 54: 323–46. https://doi.org/0.1146/annurev-phyto-080614-115913.

Hedrick, Philip W. 2005. “A Standardized Genetic Differentiation Measure.” Evolution 59 (8): 1633–8. http://dx.doi.org/10.1111/j.0014-3820.2005.tb01814.x.

Jombart, Thibaut. 2008. “adegenet: A R Package for the Multivariate Analysis of Genetic Markers.” Bioinformatics 24 (11): 1403–5. https://doi.org/10.1093/bioinformatics/btn129.

Jost, Lou. 2008. “GST And Its Relatives Do Not Measure Differentiation.” Molecular Ecology 17 (18): 4015–26. http://dx.doi.org/10.1111/j.1365-294X.2008.03887.x.

Kamvar, Zhian N, Jonah C Brooks, and Niklaus J Grünwald. 2015. “Novel R tools for analysis of genome-wide population genetic data with emphasis on clonality.” Name: Frontiers in Genetics 6: 208. https://doi.org/10.3389/fgene.2015.00208.

Kamvar, Z N, J F Tabima, and Niklaus J Grünwald. 2014. “Poppr: An R Package for Genetic Analysis of Populations with Clonal, Partially Clonal, and/or Sexual Reproduction.” PeerJ 2: e281. https://doi.org/10.7717/peerj.281.

Knaus, Brian J, and Niklaus J Grünwald. 2017. “Vcfr: A Package to Manipulate and Visualize Variant Call Format Data in R.” Molecular Ecology Resources 17 (1): 44–53. http://dx.doi.org/10.1111/1755-0998.12549.

Luikart, Gordon, Phillip R England, David Tallmon, Steve Jordan, and Pierre Taberlet. 2003. “The Power and Promise of Population Genomics: From Genotyping to Genome Typing.” Nature Reviews Genetics 4 (12): 981–94. https://doi.org/10.1038/nrg1226.

Nei, Masatoshi. 1973. “Analysis of Gene Diversity in Subdivided Populations.” Proceedings of the National Academy of Sciences 70 (12): 3321–3. http://www.pnas.org/content/70/12/3321.abstract.

Paradis, Emmanuel, Julien Claude, and Korbinian Strimmer. 2004. “APE: Analyses of Phylogenetics and Evolution in R Language.” Bioinformatics 20 (2): 289–90. https://doi.org/10.1093/bioinformatics/btg412.

Paradis, Emmanuel, Thierry Gosselin, Niklaus J Grünwald, Thibaut Jombart, Stéphanie Manel, and Hilmar Lapp. 2017. “Towards an Integrated Ecosystem of R Packages for the Analysis of Population Genetic Data.” Molecular Ecology Resources 17 (1): 1–4. https://doi.org/10.1111/1755-0998.12636.

Simpson, Edward H. 1949. “Measurement of Diversity.” Nature 163: 688. http://dx.doi.org/10.1038/163688a0.

Wright, Sewall. 1949. “The Genetical Structure of Populations.” Annals of Eugenics 15 (1): 323–54. https://doi.org/10.1111/j.1469-1809.1949.tb02451.x.